src/db/gc/mod.rs

```diff
@@ -0,0 +1,372 @@
+    // 3. Generate `depth` blocks.
+    // 4. Run the GC again to make sure it marked all the available records successfully.
+    async fn test_populate() {
+        let interval = Duration::from_secs(0);
+        let db = Arc::new(MemoryDB::default());
+        let chain_config = Arc::new(ChainConfig::default());
+
+        // Generate genesis block.
+        let gen_block: BlockHeader = mock_block(1, 1);
+        let depth = 1;
+        db.put_cbor_default(&gen_block).unwrap();
+        let cs = Arc::new(
+            ChainStore::new(db.clone(), db.clone(), chain_config, gen_block.clone()).unwrap(),
+        );
+
+        let cs_cloned = cs.clone();
+        let get_heaviest_tipset = Box::new(move || cs_cloned.heaviest_tipset());
+        let mut gc = MarkAndSweep::new(db.clone(), get_heaviest_tipset, depth, interval);
+
+        // test insufficient epochs
+        gc.gc_workflow(interval).await.unwrap();
+        assert!(gc.marked.is_empty());
+
+        // test marked
+        run_to_epoch(db, cs, depth);
+        gc.gc_workflow(interval).await.unwrap();
+        assert_eq!(gc.marked.len(), 2);
+        assert_eq!(gc.epoch_marked, 1);
+    }
+
+    #[tokio::test]
+    async fn test_filter_and_sweep() {
+        let interval = Duration::from_secs(0);
+        let db = Arc::new(MemoryDB::default());
+        let chain_config = Arc::new(ChainConfig::default());
+        let gen_block: BlockHeader = mock_block(1, 1);
+        let depth = 1;
+        db.put_cbor_default(&gen_block).unwrap();
+        let cs = Arc::new(
+            ChainStore::new(db.clone(), db.clone(), chain_config, gen_block.clone()).unwrap(),
+        );
+        let cs_cloned = cs.clone();
+        let get_heaviest_tipset = Box::new(move || cs_cloned.heaviest_tipset());
+        let mut gc = MarkAndSweep::new(db.clone(), get_heaviest_tipset, depth, interval);
+
+        run_to_epoch(db.clone(), cs.clone(), depth);
+
+        let mut reachable_cnt = (depth + 1) as u64;
+
+        let unreachable_cnt = 4;
+        // test insufficient epochs for filter step
+        insert_unreachable(db.clone(), unreachable_cnt);
+        gc.gc_workflow(interval).await.unwrap();
+        assert_eq!(gc.marked.len() as u64, reachable_cnt + unreachable_cnt);
+        assert_eq!(gc.epoch_marked, 1);
+
+        // filter and sweep
+        run_to_epoch(db.clone(), cs.clone(), depth * 2);
+        reachable_cnt += depth as u64;
+
+        assert_eq!(
+            db.get_keys().unwrap().len() as u64,
+            unreachable_cnt + reachable_cnt
+        );
+        gc.gc_workflow(interval).await.unwrap();
+        assert_eq!(gc.marked.len(), 0);
+        assert_eq!(db.get_keys().unwrap().len(), reachable_cnt as usize);
+
+        // try another run
+        gc.gc_workflow(interval).await.unwrap();
+        assert_eq!(gc.marked.len(), db.get_keys().unwrap().len());
+    }
+
+    #[quickcheck_async::tokio]
+    async fn test_workflow(depth: u8, current_epoch: u8, unreachable_cnt: u8) {
```

Comment:
This test is too long for me to understand. What properties are you testing? Are you testing more than one property? If so, can they be tested separately?

Ideally, quickcheck tests would map almost directly to properties or statements that we believe to be true. For example, we believe that no state trees within `depth` epochs will be garbage collected. We believe none of the block headers reachable from HEAD will _ever_ be garbage collected. We believe that block headers NOT reachable from HEAD will be garbage collected _after_ `chain_finality` epochs. These are invariants that should be true and it would be great to test them separately. Note, testing isn't primarily to find bugs in your code. The primary purpose of testing is to convince developers (others and yourself) that your believes about your code are correct.

Reply:
Right, that makes sense. Let's test them separately. Yeah, the test is a bit long, but I have an idea on how to simplify it.